### PR TITLE
Issue #6238: Upgrade Powermock2 to RC4

### PR DESCRIPTION
**Pull request by pbludov**
Issue #6238

Part of #6238 It simply fixed one test:
```
when(classLoader.getURLs()).thenThrow(IOException.class);
```
is not a legal code, since `URLClassLoader.getURLs()` throws nothing.

### CODE REVIEW

**Comment by pbludov** [comment]
>  Can you explain what behavior you mean?
The failing tests are `com.puppycrawl.tools.checkstyle.PackageObjectFactoryTest.RegularTests#testCreateObjectFromFullModuleNamesWithAmbiguousException` and `com.puppycrawl.tools.checkstyle.PackageObjectFactoryTest.RegularTests#testCreateObjectFromMap`

After some investigation I've got the failing line:
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/ModuleReflectionUtil.java#L91
```
return AutomaticBean.class.isAssignableFrom(clazz) // false with @RunWith(PowermockRunner.class)
```
This gives me an idea how to rewrite this test in the right way. Thank you, @rnveach

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 1]
```diff
@@ -49,18 +50,27 @@
...
 import com.puppycrawl.tools.checkstyle.api.FileText;
 import com.puppycrawl.tools.checkstyle.api.LocalizedMessage;
 import com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationLocationCheck;
 import com.puppycrawl.tools.checkstyle.internal.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.ModuleReflectionUtil;
 
 /**
- * Enter a description of class PackageObjectFactoryTest.java.
+ * This test uses the Powermock framework to simulate an exception
+ * in the methid {@link ModuleReflectionUtil#getCheckstyleModules(Collection, ClassLoader)}.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: `methid` => `method`

**Comment by rnveach** [comment]
@romani I see nothing else wrong.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 1 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 2]
```diff
@@ -328,11 +338,21 @@ public void testCreateCheckWithPartialPackageNameByBruteForce() throws Exception
     @SuppressWarnings("unchecked")
     public void testGenerateThirdPartyNameToFullModuleNameWithException() throws Exception {
         final URLClassLoader classLoader = mock(URLClassLoader.class);
-        when(classLoader.getURLs()).thenThrow(IOException.class);
-        final Method method = factory.getClass().getDeclaredMethod(
+        final Set<String> packages = new LinkedHashSet<>(1);
+        packages.add(BASE_PACKAGE);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: Please do it by stream https://stackoverflow.com/questions/2041778/how-to-initialize-hashset-values-by-construction, ones we jump to next java, it will be easy to update code, I think one inspection will hint this to update

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: Done with `Collections.singleton`

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 3]
```diff
@@ -397,6 +416,11 @@ public void testGetShortFromFullModuleNamesThirdParty() {
                 PackageObjectFactory.getShortFromFullModuleNames(fullName));
     }
 
+    /**
+     * Test check for {@link PackageObjectFactoryTest#testConstructorFailure()}.
+     *
+     * @noinspection JUnitTestCaseWithNoTests
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: We try to keep reason of suppression each time close to suppression.

Please explain reason of suppression in javadoc description.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 4]
```diff
@@ -306,7 +305,7 @@ private Object createObjectFromFullModuleNames(String name, Set<String> fullModu
                         }));
         }
         catch (IOException ignore) {
-            returnValue = new HashMap<>();
+            returnValue = Collections.emptyMap();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: No mix for functional update and minor refactoring.
Please move to separate commit of the same PR.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 5]
```diff
@@ -328,11 +338,20 @@ public void testCreateCheckWithPartialPackageNameByBruteForce() throws Exception
...
-        when(classLoader.getURLs()).thenThrow(IOException.class);
-        final Method method = factory.getClass().getDeclaredMethod(
+        final Set<String> packages = Collections.singleton(BASE_PACKAGE);
+
+        // Set up an exception in the ModuleReflectionUtil.getCheckstyleModules()
+        mockStatic(ModuleReflectionUtil.class);
+        doThrow(new IOException("mock exception")).when(ModuleReflectionUtil.class);
+        ModuleReflectionUtil.getCheckstyleModules(packages, classLoader);
+
+        // Invoke the method via reflection
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: Please do me favor and place javadoc comment over the method to briefly explain why reflection is the only way to test it.  We try to keep it as rule now, and this method was created before to agreed on it. We try to make it clear that usage of reflection is allowed but not easy way for engineer.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 3 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: done

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 5 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java (hunk 4 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: All changes to `PackageObjectFactory` were reverted.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 3 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: done

**Comment by rnveach** [comment]
@pbludov violations in teamcity are correct. Please correct them.

**Comment by pbludov** [comment]
> Please move to separate commit of the same PR, in commit, no reference to issue.
Done. All changes to `PackageObjectFactory` are here: 
https://github.com/checkstyle/checkstyle/pull/6239/commits/37c0adbebdfd1ac162f4d277ef3e1530d60ee809
